[collective] Add a torch.compile + functional_collectives test

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #109647
* __->__ #110688

Add a test to ensure functional_collectives + torch.compile always works.

Differential Revision: [D50001491](https://our.internmc.facebook.com/intern/diff/D50001491/)

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

fegin committed:
[collective] Add a torch.compile + functional_collectives test

Add a test to ensure functional_collectives + torch.compile always works.

Differential Revision: [D50001491](https://our.internmc.facebook.com/intern/diff/D50001491/)

[ghstack-poisoned]

fegin committed:
Update on "[collective] Add a torch.compile + functional_collectives test"


Add a test to ensure functional_collectives + torch.compile always works.

Differential Revision: [D50001491](https://our.internmc.facebook.com/intern/diff/D50001491/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

wanchaol commented on test/distributed/test_functional_api.py:
```diff
@@ -458,6 +458,16 @@ def test_all_gather_into_tensor_coalesced(self):
+    @skip_if_lt_x_gpu(WORLD_SIZE)
+    @requires_nccl()
+    @with_comms()
+    def test_tracing(self):
```
I am a bit surprised this issue was not captured in CI, we have bunch of tests should guard it https://github.com/pytorch/pytorch/blob/main/test/distributed/test_inductor_collectives.py#L572

but thanks for adding more tests to guard it!

wanchaol reviewed (approved):
Let's try to land this to temporarily unblock, and @yanboliang can address this fundamentally in a separate PR

fegin committed:
Update on "[collective] Add a torch.compile + functional_collectives test"


Add a test to ensure functional_collectives + torch.compile always works.

Differential Revision: [D50001491](https://our.internmc.facebook.com/intern/diff/D50001491/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

fegin committed:
Update on "[collective] Add a torch.compile + functional_collectives test"


Add a test to ensure functional_collectives + torch.compile always works.

Differential Revision: [D50001491](https://our.internmc.facebook.com/intern/diff/D50001491/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

fegin commented:
@pytorchbot  merge

fduwjj commented on test/distributed/test_functional_api.py:
```diff
@@ -544,6 +564,17 @@ def test_all_to_all_single_split_sizes_none(self):
         self.assertEqual(y, expected)
 
+    @unittest.skipIf(not has_triton(), "Inductor+gpu needs triton and recent GPU arch")
+    @skip_if_lt_x_gpu(WORLD_SIZE)
```
Well this may until run for the periodic label enabled.

fegin replied:
WORLD_SIZE is 2, so I guess it is fine?

fduwjj replied:
Oh, NVM, I thought it was 4.

fduwjj reviewed (approved):
LGTM
